morpho-checkpoint-node/src/Morpho/Config/Types.hs

```diff
@@ -116,43 +118,144 @@ deriving instance AllBF Eq f (NodeConfiguration_ Bare) => Eq (NodeConfiguration_
 
 deriving instance AllBF Show f (NodeConfiguration_ Bare) => Show (NodeConfiguration_ Bare f)
 
-parseConfigJSON :: Object -> NodeConfigurationFunctor Parser
-parseConfigJSON v =
+-- | Determines the mapping from configuration file field to 'NodeConfiguration'
```

Comment:
It's really cool you got it to work by splitting the parsing logic into field name and parser (though it is a bit less elegant when looking at this file, sadly.)

I'm curious how you got to this stage and what else you tried; for instance did you try something like a tuple of a field name + the actual parser?

Reply:
I have been thinking that it might be nicer to do the whole config in a single definition, something like

```haskell
data ConfigSources a = ConfigSources
  { fieldName :: Text
  , parseField :: Value -> Result a
  , cliParser :: Parser a
  , default :: Maybe a
  }
  deriving Functor

configSources :: NodeConfigurationFunctor ConfigSources
configSources = NodeConfiguration
  { ncProtocol = ConfigSources
    { fieldName = "Protocol"
    , parseField = fromJSON
    , cliParser = ...
    , default = Nothing
    }
  , ...
  }
```

I think that would actually be nicer. I'm thinking this could also be a future change though.